File: fries-leetcode-training/16-input-output/04-linked-list-input-output.py
"""
04-链表输入输出处理

题目描述：
演示链表的输入输出处理，包括链表构建和遍历。

输入格式：
第一行：整数n（链表长度）
第二行：n个整数，表示链表节点的值

输出格式：
输出链表的所有节点值，空格分隔

示例：
输入：
5
1 2 3 4 5

输出：
1 2 3 4 5
"""

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def build_linked_list(values):
    """根据数组构建链表"""
    if not values:
        return None
    
    head = ListNode(values[0])
    current = head
    
    for i in range(1, len(values)):
        current.next = ListNode(values[i])
        current = current.next
    
    return head

def print_linked_list(head):
    """打印链表"""
    values = []
    current = head
    
    while current:
        values.append(current.val)
        current = current.next
    
    print(*values)
    return values

def linked_list_to_array(head):
    """链表转数组"""
    result = []
    current = head
    
    while current:
        result.append(current.val)
        current = current.next
    
    return result

def test_cases():
    """测试用例"""
    print("=== 链表输入输出测试 ===")
    
    # 模拟输入
    values = [1, 2, 3, 4, 5]
    print(f"输入数组: {values}")
    
    # 构建链表
    head = build_linked_list(values)
    
    # 输出链表
    print("输出链表:")
    print_linked_list(head)
    
    # 验证转换
    converted = linked_list_to_array(head)
    print(f"转换回数组: {converted}")
    
    # 测试空链表
    print("\n空链表测试:")
    empty_head = build_linked_list([])
    print_linked_list(empty_head)

if __name__ == "__main__":
    # 运行测试
    test_cases()
    
    # 交互式输入（取消注释以启用）
    # n = int(input())
    # values = list(map(int, input().split()))
    # head = build_linked_list(values)
    # print_linked_list(head)

File: fries-leetcode-training/16-input-output/04-linked-list-input-output.cpp
/*
04-链表输入输出处理

题目描述：
演示链表的输入输出处理，包括链表构建和遍历。

输入格式：
第一行：整数n（链表长度）
第二行：n个整数，表示链表节点的值

输出格式：
输出链表的所有节点值，空格分隔

示例：
输入：
5
1 2 3 4 5

输出：
1 2 3 4 5
*/

#include <iostream>
#include <vector>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode* next) : val(x), next(next) {}
};

ListNode* buildLinkedList(const vector<int>& values) {
    if (values.empty()) return nullptr;
    
    ListNode* head = new ListNode(values[0]);
    ListNode* current = head;
    
    for (int i = 1; i < values.size(); i++) {
        current->next = new ListNode(values[i]);
        current = current->next;
    }
    
    return head;
}

void printLinkedList(ListNode* head) {
    vector<int> values;
    ListNode* current = head;
    
    while (current) {
        values.push_back(current->val);
        current = current->next;
    }
    
    for (int i = 0; i < values.size(); i++) {
        cout << values[i];
        if (i < values.size() - 1) cout << " ";
    }
    cout << endl;
}

vector<int> linkedListToArray(ListNode* head) {
    vector<int> result;
    ListNode* current = head;
    
    while (current) {
        result.push_back(current->val);
        current = current->next;
    }
    
    return result;
}

void deleteLinkedList(ListNode* head) {
    while (head) {
        ListNode* temp = head;
        head = head->next;
        delete temp;
    }
}

void testCases() {
    cout << "=== 链表输入输出测试 ===" << endl;
    
    // 模拟输入
    vector<int> values = {1, 2, 3, 4, 5};
    cout << "输入数组: ";
    for (int i = 0; i < values.size(); i++) {
        cout << values[i];
        if (i < values.size() - 1) cout << " ";
    }
    cout << endl;
    
    // 构建链表
    ListNode* head = buildLinkedList(values);
    
    // 输出链表
    cout << "输出链表:" << endl;
    printLinkedList(head);
    
    // 验证转换
    vector<int> converted = linkedListToArray(head);
    cout << "转换回数组: ";
    for (int i = 0; i < converted.size(); i++) {
        cout << converted[i];
        if (i < converted.size() - 1) cout << " ";
    }
    cout << endl;
    
    // 测试空链表
    cout << "\n空链表测试:" << endl;
    ListNode* emptyHead = buildLinkedList({});
    printLinkedList(emptyHead);
    
    // 清理内存
    deleteLinkedList(head);
    deleteLinkedList(emptyHead);
}

int main() {
    // 运行测试
    testCases();
    
    // 交互式输入（取消注释以启用）
    // int n;
    // cin >> n;
    // vector<int> values(n);
    // for (int i = 0; i < n; i++) {
    //     cin >> values[i];
    // }
    // ListNode* head = buildLinkedList(values);
    // printLinkedList(head);
    // deleteLinkedList(head);
    
    return 0;
}
